Feature : Visualizer (Waveform, Spectrum, Equalizer)

bzawad opened the pull request:
This adds an optional visualizer on the OLED screen when the UV level meter is enabled.

The visualizer can be enabled under SETTINGS / COMMUNITY FEATURES / VISUALIZER by selecting a type.  The default is OFF.  The three types are...

WAVEFORM - A oscilloscope style waveform visualizer.
SPECTRUM - A frequency graph visualizer.
EQUALIZER - A 16 bar graphic equalizer visualizer with delayed peak indicators.

When a visualizer is enabled then it will be displayed when the UV meter is enabled in Song or Session mode by pressing Affect Entire and then Level.

When the visualizer is active (displaying on screen), you can switch between modes using the following hot keys:

- :key[SYNTH] - Switch to Waveform mode
- :key[KIT] - Switch to Spectrum mode
- :key[MIDI] - Switch to Equalizer mode

These mode changes are temporary for the current session and don't affect your community setting preference.

You can also toggle the visualizer on or off independently using :key[Shift + Level]. This provides a quick way to enable the visualizer without needing to set up the VU meter conditions, and shows popup feedback indicating the current state.

When the visualizer is active certain messages now use popup overlays instead of direct canvas text to avoid rendering conflicts:
- Tempo changes: “Tempo: XXX.X”
- Key/scale changes: new key/scale name
- Bars/Beats remaining: countdown overlays
- Resampling notices: “Resampling will end…”

Behavior:
- Visualizer off → original direct text
- Visualizer on and active → popup overlays that auto-clear

Benefits:
- Text remains visible above visualizer
- Popups clear correctly
- No impact on non-visualizer users
- Preserves all previous behavior

Discord feature thread: https://discord.com/channels/608916579421257728/1435862425864896542

**Photos**
<img width="800" height="1067" alt="IMG_6475" src="https://github.com/user-attachments/assets/265b2979-1d48-413f-9ec9-dcc139e08982" />
<img width="800" height="1067" alt="IMG_6476" src="https://github.com/user-attachments/assets/e44209ee-f226-496e-a5f1-98ddcad5d482" />
<img width="800" height="1067" alt="IMG_6469" src="https://github.com/user-attachments/assets/d8aa8070-87fe-44f7-a6ac-647b60ff5a1d" />
<img width="800" height="1067" alt="IMG_6470" src="https://github.com/user-attachments/assets/8193ef31-975f-4a0d-820d-34b5b435d461" />
<img width="800" height="1067" alt="IMG_6472" src="https://github.com/user-attachments/assets/04338df8-f1a9-42b8-82b8-6f5d38f8ae89" />

<img width="3024" height="4032" alt="image" src="https://github.com/user-attachments/assets/a28989df-2cf2-41c1-8c81-c6a25f989642" />

seangoodvibes commented on src/deluge/gui/views/view.cpp:
```diff
@@ -1839,6 +1878,870 @@ void View::renderVUMeter(int32_t maxYDisplay, int32_t xDisplay, RGB thisImage[][
 	}
 }
 
+// Static buffers for spectrum visualizer FFT computation
```
I don't think all of this should be in the view class

I think it would be cleaner if this got refactored out into a separate visualizer class which inherits from the OLED class I think.

bzawad replied:
I split it all out.  There's a new visualizer and separate files for each of the visualizer types.

bzawad commented:
I found I can fix the problematic text such as key changes, bars remaining, and temp change by having them use the Pop System instead of the direct canvas write which get overwritten by the visualizer. However, I'm not sure what the UX plan is for those or why they use direct canvas write.  The popups look nice!

<img width="3024" height="4032" alt="image" src="https://github.com/user-attachments/assets/ad06067b-29e4-4795-90f5-788c80637e64" />

bzawad commented:
@seangoodvibes I reorganized the code so everything is cleanly under the visualizer domain now.  

Is there anything I need to do about those different clang naming issues?  It looks like a lot of the warnings were for variables that seemed to be following the same convention as other variables in the same area.

Also, is there anything I should do about the interference with text such as tempo change?  I found those are writing directly to the canvas and switching to popup system fixes those.  I can leave those for someone else if there's a plan for moving to popups or something.

seangoodvibes commented:
> @seangoodvibes I reorganized the code so everything is cleanly under the visualizer domain now.  
> 
> 
> 
> Is there anything I need to do about those different clang naming issues?  It looks like a lot of the warnings were for variables that seemed to be following the same convention as other variables in the same area.
> 
> 
> 
> Also, is there anything I should do about the interference with text such as tempo change?  I found those are writing directly to the canvas and switching to popup system fixes those.  I can leave those for someone else if there's a plan for moving to popups or something.

Thanks!

Re clang: basically that clang formatting was introduced long after the codebase went open source.

Basically the idea now is that anything new should follow the clang formatting rules (eg get rid of all warnings).

Over time the older code should get cleaned up too but right now priority is just the warnings for the changes in your PR.

As for the popups, those things were built on the assumption that they could render to the display and not interfere with anything else. Obviously in this case it does not make sense to try and overwrite your canvas, so a popup makes sense if the visualizer is active.

So I think a check of if -> visualizer, do popup, else render to canvas would make sense to me.

seangoodvibes commented on src/deluge/playback/playback_handler.cpp:
```diff
@@ -2324,15 +2324,32 @@ void PlaybackHandler::displayTempoBPM(float tempoBPM) {
+		bool isSessionOrArranger = (currentUI == &sessionView || currentUI == &arrangerView);
+
+		// Check if visualizer is enabled AND actively running
+		uint32_t visualizer_mode = runtimeFeatureSettings.get(RuntimeFeatureSettingType::Visualizer);
```
I think I've seen this kind of check a few times, maybe you could refactor it out into the visualizer class so you can just visualizer::isEnabled()

seangoodvibes commented on src/deluge/playback/playback_handler.cpp:
```diff
@@ -2324,15 +2324,32 @@ void PlaybackHandler::displayTempoBPM(float tempoBPM) {
+		                          || (visualizer_mode == RuntimeFeatureStateVisualizer::VisualizerEqualizer);
+
+		bool visualizer_active = false;
+		if (visualizer_enabled) {
```
Maybe this could could be factored out too so you can just do visualizer::isActive()

seangoodvibes commented on src/deluge/gui/views/arranger_view.cpp:
```diff
@@ -3091,6 +3112,20 @@ void ArrangerView::graphicsRoutine() {
 
+	// Request OLED refresh for visualizer if active (ensures continuous updates)
+	uint32_t visualizer_mode = runtimeFeatureSettings.get(RuntimeFeatureSettingType::Visualizer);
+	bool visualizer_enabled = (visualizer_mode == RuntimeFeatureStateVisualizer::VisualizerWaveform)
```
Same comment as above

seangoodvibes commented on src/deluge/gui/views/session_view.cpp:
```diff
@@ -1900,6 +1901,24 @@ void SessionView::renderOLED(deluge::hid::display::oled_canvas::Canvas& canvas)
 
+	// Check if visualizer should be displayed (same conditions as VU meter)
+	uint32_t visualizer_mode = runtimeFeatureSettings.get(RuntimeFeatureSettingType::Visualizer);
+	bool visualizer_enabled = (visualizer_mode == RuntimeFeatureStateVisualizer::VisualizerWaveform)
```
Same comment as above

seangoodvibes commented on src/deluge/gui/views/session_view.cpp:
```diff
@@ -1913,15 +1932,104 @@ void SessionView::renderOLED(deluge::hid::display::oled_canvas::Canvas& canvas)
+
+				// Check if visualizer is enabled AND actively running
+				uint32_t visualizer_mode = runtimeFeatureSettings.get(RuntimeFeatureSettingType::Visualizer);
+				bool visualizer_enabled = (visualizer_mode == RuntimeFeatureStateVisualizer::VisualizerWaveform)
```
Same comment as above

bzawad commented on src/deluge/gui/views/session_view.cpp:
```diff
@@ -1913,15 +1932,104 @@ void SessionView::renderOLED(deluge::hid::display::oled_canvas::Canvas& canvas)
+				                          || (visualizer_mode == RuntimeFeatureStateVisualizer::VisualizerEqualizer);
+
+				bool visualizer_active = false;
+				if (visualizer_enabled) {
```
@seangoodvibes I got all of those above.

seangoodvibes commented on src/deluge/hid/display/visualizer.cpp:
```diff
@@ -0,0 +1,220 @@
+	return false;
+}
+
+void Visualizer::requestVisualizerUpdateIfNeeded(View* view) {
```
is passing &view as a pointer necessary?

I think you can simply do view.activeModControllable, view.displayVUMeter, view.activeModControllableModelStack instead.

bzawad commented:
I think the code is in a pretty solid state but I'd like to get feedback from users/testers on the actual implementations of the visualizers where I've made my own judgements.  Examples are the default amplitude level, the low/high frequency biases for the spectrum and EQ, etc.

seangoodvibes commented on src/deluge/hid/display/visualizer.cpp:
```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright © 2014-2023 Synthstrom Audible Limited
```
You can put your name here for the copyright as this is your work :) 

(c) 2025 your name

bzawad commented:
> > Ok got the last few things. Thanks for all the help. Looks much better now. All clean new green code.  Only 1,671 lines! 😆
> 
> Looks great! Thanks for bearing with me with all the feedback hahah.
> 
> If you would like user feedback / testing, I think you can create a feature thread on the discord and attach the firmware bin file for folks to try it out :)
> 
> Cheers!

@seangoodvibes Should I rebase to the latests before doing this?  I'm not really sure how stable things are.  If so I'll do this after I get the last of these clang issues resolved.  Just pushed another cleanup.

bzawad commented:
Discord feature thread: https://discord.com/channels/608916579421257728/1435862425864896542

seangoodvibes commented:
@bzawad could you document this feature in the website features page? 

you can refer to the other feature documentation for reference: https://delugecommunity.com/features/arpeggiator/

bzawad commented:
I made some tuning changes and overall the appearance is better but I'm really struggling with issues with frequency detection of the spectrum and equalizer.  31Hz starting too high like the range is about 5 octaves off and I can't figure out why.  When I apply multipliers then that can adjust it down but then frequency starts to sweep incorrectly above 1 khz.

It looks a lot better but incorrect value aren't really helpful!  I'm kind of stuck on this.

I'll update docs soon after the dust settles.

We have hotkeys now, SYNTH, KIT, and MIDI change modes.  Open to any other suggestions for hotkeys.

bzawad commented:
I added a new hotkey combination (Shift+Level) to view the visualizer separately from the UV meter.  It's just another option and everything else still works the same.  I updated documentation and addressed the clang-tidy issues that I saw related to my code.  All of the visualizers seem to be working pretty good.  I think at this point it's functionally complete unless we want to change hotkeys or there are some bugs I haven't found yet.  I'll post an updated firmware in the forum.

seangoodvibes commented on website/src/content/docs/features/visualizer.mdx:
```diff
@@ -0,0 +1,100 @@
+- **Sound Design**: The Spectrum mode provides insight into the harmonic content of your sounds
+- **Troubleshooting**: The Waveform mode can help identify clipping or level issues
+- **Quick Mode Switching**: Use the hot keys to quickly compare different visualization perspectives
+- **Independent Access**: Use :key[Shift + Level] to quickly toggle the visualizer on or off without setting up VU meter conditions, perfect for quick visual checks during production
```
```suggestion
- **Independent Access**: Use :key[Shift] + `LEVEL/PAN` mod button to quickly toggle the visualizer on or off without setting up VU meter conditions, perfect for quick visual checks during production
```

seangoodvibes commented on website/src/content/docs/features/visualizer.mdx:
```diff
@@ -0,0 +1,100 @@
+
+The visualizer will automatically appear in the OLED display area when these conditions are satisfied, replacing the standard VU meter display. The VU meter will stop displaying (and therefore the visualizer will be hidden) if you switch mod button selections, turn :key[AffectEntire] off, select a clip, or exit Song/Arranger views.
+
+Alternatively, you can use the independent visualizer toggle to display the visualizer without needing to enable the VU meter. Press :key[Shift + Level] to toggle the visualizer on or off independently. This allows you to view visualizations in situations where the VU meter conditions cannot be met.
```
```suggestion
Alternatively, you can use the independent visualizer toggle to display the visualizer without needing to enable the VU meter. Press :key[Shift] + `LEVEL/PAN` mod button to toggle the visualizer on or off independently. This allows you to view visualizations in situations where the VU meter conditions cannot be met.
```

bzawad commented:
Ok I fixed whatever happened and got things back to normal.  Rebased and squashed and added the additional documentation.

seangoodvibes commented:
> @seangoodvibes I have a couple branches with the improved visualizer. I think I would like to replace the version in this PR with the new improved version. The main reason besides the additional visualizers is that the new version has improved naming and organization for the standard visualizers, for example, Spectrum -> Line Spectrum and Equalizer -> Bar Spectrum and then there's the MIDI Piano Roll so the hotkeys change some.
> 
> I can either force push the 1.4 version into this branch or close this PR out and create a new PR. Do you have any preference?
> 
> Improved Visualizer 1.4 branch... https://github.com/bzawad/DelugeFirmware/tree/feature/oscilloscope-visualizer-more
> 
> This version would still have the other community bugs until those are sorted.
> 
> I also have this 1.3 branch... https://github.com/bzawad/DelugeFirmware/tree/feature/oscilloscope-visualizer-more-1.3
> 
> Is there anything you recommend doing with this? The main purpose of it is just a stable version for people to use now.
> 
> Thanks!

Hey, feel free to close this PR and open a new one based off 1.4.

I don't really have any suggestions regarding 1.3 other than I think it's a good idea to have a 1.3 branch of the feature for testing just because there are some gremlins / bugs still in 1.4 and it will help rule out that your feature is introducing anything.

Cheers!

bzawad commented:
Closing this PR to replace with https://github.com/SynthstromAudible/DelugeFirmware/pull/4214 for the improved version.
